### DISCUSSION TITLE
How to run `git clone` to a self-hosted Gitlab instance behind Teleport Application Access

### DISCUSSION THREAD

**Discussion by pschisa** [gravitational/teleport#16162] [Show and tell]
Teleport can be used with [Application Access](https://goteleport.com/docs/application-access/introduction/) to securely provide access to self-hosted applications such as Gitlab.  See the options below for how to still use `git clone` when your Gitlab instance is protected behind Teleport Application Access.  This discussion assumes the Gitlab application is already registered with Teleport.  Review our [dynamic registration guide](https://goteleport.com/docs/application-access/guides/dynamic-registration/) to register your Gitlab application with Teleport otherwise.

### Method 1: Use the application certificates

1. Login to the Teleport cluster (`tsh login proxy.example.com`)
2. Login to the Gitlab application through Teleport (`tsh app login <gitlab-app-name>`).  The output from this command will contain an example `curl` command and list the certificate, key, and CA.
3. Run your `git clone` command using the credentials provided by Teleport.  Be sure to adjust your repo destination to `https`

Example:
~~~
$ tsh login test-cluster1.schisa.cf
If browser window does not open automatically, open it by clicking on the link:
 http://127.0.0.1:51297/1bcb1008-e58f-4068-a04f-45ea7618c448
> Profile URL:        https://test-cluster1.schisa.cf:443
  Logged in as:       pschisa
  Cluster:            ip-172-31-36-239-ec2-internal
  Roles:              admin2
  Logins:             pschisa, root, ubuntu, paulschisa, ec2-user, -teleport-internal-join
  Kubernetes:         enabled
  Kubernetes cluster: "kate1-test"
  Kubernetes groups:  system:masters
  Valid until:        2022-09-06 13:33:52 -0400 EDT [valid for 1h0m0s]
  Extensions:         permit-agent-forwarding, permit-port-forwarding, permit-pty

$ tsh app login gitlab-fips
Logged into app gitlab-fips. Example curl command:

curl \
  --cacert /Users/paulschisa/.tsh/keys/test-cluster1.schisa.cf/cas/ip-172-31-36-239-ec2-internal.pem \
  --cert /Users/paulschisa/.tsh/keys/test-cluster1.schisa.cf/pschisa-app/ip-172-31-36-239-ec2-internal/gitlab-fips-x509.pem \
  --key /Users/paulschisa/.tsh/keys/test-cluster1.schisa.cf/pschisa \
  https://gitlab-fips.test-cluster1.schisa.cf

$ git -c http.sslCert=/Users/paulschisa/.tsh/keys/test-cluster1.schisa.cf/pschisa-app/ip-172-31-36-239-ec2-internal/gitlab-fips-x509.pem -c http.sslKey=/Users/paulschisa/.tsh/keys/test-cluster1.schisa.cf/pschisa -c http.sslCAPath=/Users/paulschisa/.tsh/keys/test-cluster1.schisa.cf/cas/ip-172-31-36-239-ec2-internal.pem clone https://gitlab-fips.test-cluster1.schisa.cf/gitlab-instance-eac530b1/test-clone.git
Cloning into 'test-clone'...
remote: Enumerating objects: 3, done.
remote: Counting objects: 100% (3/3), done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 3 (delta 0), reused 0 (delta 0), pack-reused 0
Receiving objects: 100% (3/3), done.
~~~

### Method 2: Use the `tsh proxy app` functionality

1. Login to the Teleport cluster (`tsh login proxy.example.com`)
2. Login to the Gitlab application through Teleport (`tsh app login <gitlab-app-name>`).
3. Run the command `tsh proxy app --port=54321 <app-name>` to open a proxy session to a `127.0.0.1:54321`
4. In a separate terminal, attempt the git clone using HTTP and the local address/port

Example:
~~~
$ tsh login test-cluster1.schisa.cf
If browser window does not open automatically, open it by clicking on the link:
 http://127.0.0.1:51297/1bcb1008-e58f-4068-a04f-45ea7618c448
> Profile URL:        https://test-cluster1.schisa.cf:443
  Logged in as:       pschisa
  Cluster:            ip-172-31-36-239-ec2-internal
  Roles:              admin2
  Logins:             pschisa, root, ubuntu, paulschisa, ec2-user, -teleport-internal-join
  Kubernetes:         enabled
  Kubernetes cluster: "kate1-test"
  Kubernetes groups:  system:masters
  Valid until:        2022-09-06 13:33:52 -0400 EDT [valid for 1h0m0s]
  Extensions:         permit-agent-forwarding, permit-port-forwarding, permit-pty

$ tsh app login gitlab-fips
Logged into app gitlab-fips. Example curl command:

curl \
  --cacert /Users/paulschisa/.tsh/keys/test-cluster1.schisa.cf/cas/ip-172-31-36-239-ec2-internal.pem \
  --cert /Users/paulschisa/.tsh/keys/test-cluster1.schisa.cf/pschisa-app/ip-172-31-36-239-ec2-internal/gitlab-fips-x509.pem \
  --key /Users/paulschisa/.tsh/keys/test-cluster1.schisa.cf/pschisa \
  https://gitlab-fips.test-cluster1.schisa.cf

$ tsh proxy app --port=54321 gitlab-fips
Proxying connections to gitlab-fips on 127.0.0.1:54321

#Move to another terminal

$ git clone http://127.0.0.1:54321/gitlab-instance-eac530b1/test-clone.git
Cloning into 'test-clone'...
Username for 'http://127.0.0.1:54321': root
Password for 'http://root@127.0.0.1:54321':
remote: Enumerating objects: 3, done.
remote: Counting objects: 100% (3/3), done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 3 (delta 0), reused 0 (delta 0), pack-reused 0
Receiving objects: 100% (3/3), done.
~~~

**Comment by dfang** [comment]
hi @pschisa, thank you for your guide,  both works, but they are complicated. not easy for someone who don't use command line too much.

1. for the first method, you need to figure out three keys and then compose a complicated command. for the next pull and push, you need to set -c parameter for git, or set an `alias tgit="git -c xxxx"`

2. for the second method, you need to remember and type the password, but we can set `git config --global credential.helper store`, it's acceptable for just input password for once.

**Comment by scinel-sys** [comment]
I had this working with option-1 for older versions of Teleport. Now when I try to connect, I get curl output like below but could not use git functions.
tsh app login gitlab
Logged into app gitlab. Example curl command:

curl \
  --cert "....\\gitlab.crt" \
  --key "....gitlab.key" \
  https://myserveraddress
  
Your second method still works fine for me. Thank for the information.
